File: pset6/sentimental-credit/credit.py
# TODO
from cs50 import get_int


def main():

    # Get card number and turn into a list to iterate through it
    card_number = get_int("Number: ")
    card_type = int(str(card_number)[:2])
    card_number = list(map(int, str(card_number)))

    checksum = 0

    # Reverse list and iterate through it
    for digit in enumerate(reversed(card_number)):

        # If odd index multiply card digit and sum the two separate digits
        if digit[0] % 2 > 0:

            result = map(int, str(digit[1] * 2))
            checksum = checksum + sum(result)

        # Sum digits together
        else:
            checksum = checksum + digit[1]

    # Check to see if card is valid
    if checksum % 10 == 0 and 12 < len(card_number) < 17 and len(card_number) != 14:

        if (card_type == 34 or card_type == 37):
            print("AMEX")

        elif (40 <= card_type <= 49):
            print("VISA")

        elif (51 <= card_type <= 55):
            print("MASTERCARD")

        else:
            print("INVALID")

    else:
        print("INVALID")


if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <math.h>
#include <cs50.h>

int main(void)
{
    long long card_number;
    int card_type;
    int digit;
    int even;
    int even_sum = 0;
    int i = 0;
    int odd_sum = 0;

    card_number = get_long("Number: ");

    // The loop finds out the length of the card while doing Luhn's Algorithm
    while (card_number > 0)
    {
        i++;

        if (i % 2 == 0)
        {
            digit = card_number % 10;
            card_number /= 10;
            //printf("Card Number: %ld\n", card_number);
            even = digit * 2;

            // If the product is over 10 separate the two digits and add together
            if (even >= 10)
            {
                even_sum += even % 10;
                even_sum += even / 10;
            }
            else
            {
                even_sum += even;
            }
        }
        else
        {
            odd_sum += card_number % 10;
            card_number /= 10;
            //printf("Card Number: %ld\n", card_number);
        }

        // Probably a stupid way to catch the first two digits of the card number while doing this loop
        if (card_number <= 99 && card_number >= 10)
        {
            card_type = card_number;
        }

        //printf("Even sum: %d\n", even_sum);
        //printf("Odd sum: %d\n", odd_sum);
    }

    int checksum = (even_sum + odd_sum) % 10;

    // Checking if the checksum is valid and the card lengths are valid
    if (checksum == 0 && i > 12 && i < 17 && i != 14)
    {
        if (card_type == 34 || card_type == 37)
        {
            printf("AMEX\n");
        }
        else if (card_type >= 40 && card_type <= 49)
        {
            printf("VISA\n");
        }
        else if (card_type >= 51 && card_type <= 55)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}